crates/rpc/src/jsonrpc.rs

```diff
@@ -0,0 +1,862 @@
+            (
+                Self::ApplicationError {
+                    code: l_code,
+                    message: l_message,
+                },
+                Self::ApplicationError {
+                    code: r_code,
+                    message: r_message,
+                },
+            ) => l_code == r_code && l_message == r_message,
+            _ => core::mem::discriminant(self) == core::mem::discriminant(other),
+        }
+    }
+}
+
+impl RpcError {
+    fn code(&self) -> i32 {
+        match self {
+            RpcError::ParseError => -32700,
+            RpcError::InvalidRequest => -32600,
+            RpcError::MethodNotFound { .. } => -32601,
+            RpcError::InvalidParams => -32602,
+            RpcError::InternalError(_) => 32603,
+            RpcError::ApplicationError { code, .. } => *code,
+        }
+    }
+
+    fn message(&self) -> &str {
+        match self {
+            RpcError::ParseError => "Parse error",
+            RpcError::InvalidRequest => "Invalid Request",
+            RpcError::MethodNotFound { method } => "Method not found",
+            RpcError::InvalidParams => "Invalid params",
+            RpcError::InternalError(_) => "Internal error",
+            RpcError::ApplicationError { code, message } => message,
+        }
+    }
+}
+
+impl Serialize for RpcError {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use serde::ser::SerializeMap;
+
+        let mut obj = serializer.serialize_map(Some(2))?;
+        obj.serialize_entry("code", &self.code())?;
+        obj.serialize_entry("message", self.message())?;
+        obj.end()
+    }
+}
+
+impl Serialize for RpcResponse {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use serde::ser::SerializeMap;
+
+        let mut obj = serializer.serialize_map(Some(3))?;
+        obj.serialize_entry("jsonrpc", "2.0")?;
+
+        match &self.output {
+            Ok(x) => obj.serialize_entry("result", &x)?,
+            Err(e) => obj.serialize_entry("error", &e)?,
+        };
+
+        match &self.id {
+            RequestId::Number(x) => obj.serialize_entry("id", &x)?,
+            RequestId::String(x) => obj.serialize_entry("id", &x)?,
+            RequestId::Null => obj.serialize_entry("id", &Value::Null)?,
+        };
+
+        obj.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for RpcRequest {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        /// Replaces [Option<Value>] because serde maps both `None` and `null`to [Option::None].
+        ///
+        /// With this helper, null is correctly mapped to [IdHelper::Some(Value::Null)].
+        #[derive(Deserialize, Default)]
+        #[serde(untagged)]
+        enum IdHelper {
+            Some(Value),
```

Comment:
Please use some borrowed representation here.

----
Using `serde::Value` instead of some "borrow" representation allows for somewhat DoS-ing with arbitrarily complex invalid json rpc IDs which are valid json and trigger useless heap allocs.